src/learning/gmm.rs

```diff
@@ -326,11 +338,30 @@ impl GaussianMixtureModel {
             new_covs.push(cov_mat / sum_weights[k]);
 
         }
-
+        self.bic = self.calculate_bic(samples);
```

Comment:
Do we need to do this within each `update_params` call? The `bic` function can only be called after training anyway so maybe we should simply store the number of samples as an `Option` field in the model?

This feels a little messy but may be worth it for the (perhaps minor) performance gain.